modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt

```diff
@@ -43,42 +82,41 @@ data class DeferredK<out A>(private val deferred: Deferred<A>, val scope: Corout
       }
     }
 
-  fun continueOn(ctx: CoroutineContext): DeferredK<A> =
-    scope.asyncK(ctx, CoroutineStart.LAZY) {
-      deferred.await()
+  fun continueOn(ctx: CoroutineContext): DeferredK<A> = when (this) {
+    is Generated -> Generated(ctx, coroutineStart, scope) {
+      scope.async(this@DeferredK.ctx, coroutineStart) {
+        generator()
+      }.await()
     }
+    is Wrapped -> scope.asyncK(ctx, CoroutineStart.LAZY) { memoized.await() }
+  }
 
   override fun equals(other: Any?): Boolean =
     when (other) {
-      is DeferredK<*> -> this.deferred == other.deferred
-      is Deferred<*> -> this.deferred == other
+      is DeferredK<*> -> this.memoized == other.memoized
+      is Deferred<*> -> this.memoized == other
       else -> false
     }
 
-  override fun hashCode(): Int = deferred.hashCode()
+  override fun hashCode(): Int = memoized.hashCode()
 
   companion object {
-    fun unit(): DeferredK<Unit> =
-      CompletableDeferred(Unit).k()
-
-    fun <A> just(a: A): DeferredK<A> =
-      CompletableDeferred(a).k()
+    fun unit(): DeferredK<Unit> = just(Unit)
 
-    fun <A> defer(scope: CoroutineScope = GlobalScope, ctx: CoroutineContext = Dispatchers.Default, start: CoroutineStart = CoroutineStart.LAZY, f: suspend () -> A): DeferredK<A> =
-      scope.asyncK(ctx, start) { f() }
+    fun <A> just(a: A): DeferredK<A> = CompletableDeferred(a).k()
 
     fun <A> defer(scope: CoroutineScope = GlobalScope, ctx: CoroutineContext = Dispatchers.Default, start: CoroutineStart = CoroutineStart.LAZY, fa: () -> DeferredKOf<A>): DeferredK<A> =
-      scope.asyncK(ctx, start) { fa().await() }
+      Generated(ctx, start, scope) { fa().await() }
 
-    operator fun <A> invoke(scope: CoroutineScope = GlobalScope, ctx: CoroutineContext = Dispatchers.Default, start: CoroutineStart = CoroutineStart.LAZY, f: () -> A): DeferredK<A> =
-      scope.asyncK(ctx, start) { f() }
-
-    fun <A> failed(t: Throwable): DeferredK<A> =
-      CompletableDeferred<A>().apply { completeExceptionally(t) }.k()
+    operator fun <A> invoke(scope: CoroutineScope = GlobalScope, ctx: CoroutineContext = Dispatchers.Default, start: CoroutineStart = CoroutineStart.LAZY, f: suspend () -> A): DeferredK<A> =
+      Generated(ctx, start, scope, f)
 
     fun <A> raiseError(t: Throwable): DeferredK<A> =
       failed(t)
 
+    fun <A> failed(t: Throwable): DeferredK<A> =
```

Comment:
*squint* why is this here? why do we need an alias?

Reply:
I know I wrote it, I cannot remember why.

Reply:
Other than providing a more intuitive constructor for those that have never used raiseError I don't see the point either :) I'll remove it for now